extensions/procfs/ProcFsSerialization.h

```diff
@@ -0,0 +1,183 @@
+  gsl_Expects(duration > 0ms);
+  serializer("Bytes Received/sec", net_dev_data.getBytesReceived()/duration.count());
+  serializer("Packets Received/sec", net_dev_data.getPacketsReceived()/duration.count());
+  serializer("Receive Errors/sec", net_dev_data.getReceiveErrors()/duration.count());
+  serializer("Receive Drop Errors/sec", net_dev_data.getReceiveDropErrors()/duration.count());
+  serializer("Receive Fifo Errors/sec", net_dev_data.getReceiveFifoErrors()/duration.count());
+  serializer("Receive Frame Errors/sec", net_dev_data.getReceiveFrameErrors()/duration.count());
+  serializer("Compressed Packets Received/sec", net_dev_data.getCompressedPacketsReceived()/duration.count());
+  serializer("Multicast Frames Received/sec", net_dev_data.getMulticastFramesReceived()/duration.count());
+
+  serializer("Bytes Transmitted/sec", net_dev_data.getBytesTransmitted()/duration.count());
+  serializer("Packets Transmitted/sec", net_dev_data.getPacketsTransmitted()/duration.count());
+  serializer("Transmit errors/sec", net_dev_data.getTransmitErrors()/duration.count());
+  serializer("Transmit drop errors/sec", net_dev_data.getTransmitDropErrors()/duration.count());
+  serializer("Transmit fifo errors/sec", net_dev_data.getTransmitFifoErrors()/duration.count());
+  serializer("Transmit collisions/sec", net_dev_data.getTransmitCollisions()/duration.count());
+  serializer("Transmit carrier losses/sec", net_dev_data.getTransmitCarrierLosses()/duration.count());
+  serializer("Compressed Packets Transmitted/sec", net_dev_data.getCompressedPacketsTransmitted()/duration.count());
+}
+
+template<class Serializer>
+requires std::invocable<Serializer, const  char*, uint64_t> &&
+         std::invocable<Serializer, const char*, std::string_view>
+void SerializeProcessStat(const ProcessStat& process_stat,
+                          Serializer serializer) {
+  serializer("COMM", process_stat.getComm());
+  serializer("RES", process_stat.getMemory());
+  serializer("CPUTIME", process_stat.getCpuTime().count());
+}
+
+template<class Serializer>
+requires std::invocable<Serializer, const char*, double> &&
+         std::invocable<Serializer, const char*, uint64_t> &&
+         std::invocable<Serializer, const char*, std::string_view>
+void SerializeNormalizedProcessStat(const ProcessStat& process_stat_start,
+                                    const ProcessStat& process_stat_end,
+                                    const std::chrono::duration<double> all_cpu_time,
+                                    Serializer serializer) {
+  gsl_Expects(all_cpu_time > 0ms);
+  gsl_Expects(process_stat_start.getComm() == process_stat_end.getComm());
+  gsl_Expects(process_stat_end.getCpuTime() >= process_stat_start.getCpuTime());
+  auto cpu_time_diff = process_stat_end.getCpuTime()-process_stat_end.getCpuTime();
+  serializer("COMM", process_stat_start.getComm());
+  serializer("RES", process_stat_end.getMemory());
+  serializer("CPU%", cpu_time_diff/all_cpu_time);
```

Comment:
They are both chrono::durations so the different base is not a problem, chrono works it out in the background.

e.g. cpu_time_diff (102) / all_cpu_time (1.0216666666480403) -> 0.99836867864789169

(however i did find a serious typo in line 177 while checking this out)

Reply:
nice, I missed the typo on line 177 :smiley: 

OK, so `chrono::duration` is smarter than I thought; but then we are missing a factor of 100, aren't we?  If `cpu_time_diff` is 23 seconds (2300 ticks) and `all_cpu_time` is 100 seconds, then "CPU%" should be "23", but right now it is "0.23".

Reply:
you are right :+1:, in my mind anything relative is between 0-1, but since I named it CPU% then it should be 0-100 https://github.com/apache/nifi-minifi-cpp/pull/1152/commits/4691fa9cfff94471f19b85180e634642fad4a736